clusterDeployer.py

```diff
@@ -29,222 +29,654 @@
+            cmd = f"virsh net-update default delete ip-dhcp-range \"{host_xml}\" --live --config"
+            r = self.hostconn.run(cmd)
+            logger.debug(r.err if r.err else r.out)
+
+            host_xml = f"<range start='{new_range}' end='192.168.122.254'/>"
+            cmd = f"virsh net-update default add ip-dhcp-range \"{host_xml}\" --live --config"
+            r = self.hostconn.run(cmd)
+            logger.debug(r.err if r.err else r.out)
+
+    def _network_xml(self, ip: str, dhcp_range: Optional[Tuple[str, str]] = None) -> str:
+        if dhcp_range is None:
+            dhcp_part = ""
+        else:
+            dhcp_part = f"""<dhcp>
+                <range start='{dhcp_range[0]}' end='{dhcp_range[1]}'/>
+                </dhcp>"""
+
+        return f"""
+                <network>
+                <name>default</name>
+                <forward mode='nat'/>
+                <bridge name='virbr0' stp='off' delay='0'/>
+                <ip address='{ip}' netmask='255.255.0.0'>
+                    {dhcp_part}
+                </ip>
+                </network>"""
+
+    def configure(self, api_network: str) -> None:
+        hostname = self.hostconn.hostname()
+        cmd = "systemctl enable libvirtd"
+        self.hostconn.run_or_die(cmd)
+        cmd = "systemctl start libvirtd"
+        self.hostconn.run_or_die(cmd)
+
+        # stp must be disabled or it might conflict with default configuration of some physical switches
+        # 'bridge' section of network 'default' can't be updated => destroy and recreate
+        # check that default exists and contains stp=off
+        cmd = "virsh net-dumpxml default"
+        ret = self.hostconn.run(cmd)
+
+        if "stp='off'" not in ret.out or "range start='192.168.122.2'" in ret.out:
+            logger.info("Destoying and recreating bridge")
+            logger.info(f"creating default-net.xml on {hostname}")
+            if hostname == "localhost":
+                contents = self._network_xml('192.168.122.1', ('192.168.122.129', '192.168.122.254'))
+            else:
+                contents = self._network_xml('192.168.123.250')
+
+            bridge_xml = os.path.join("/tmp", 'vir_bridge.xml')
+            self.hostconn.write(bridge_xml, contents)
+            # Not sure why/whether this is needed. But we saw failures w/o it.
+            # Without this, net-undefine within ensure_bridge_is_started fails as libvirtd fails to restart
+            # We need to investigate how to remove the sleep to speed up
+            time.sleep(5)
+            self._ensure_started(api_network, bridge_xml)
+
+            self.limit_dhcp_range("192.168.122.2", "192.168.122.129")
+
+            cmd = "systemctl restart libvirtd"
+            self.hostconn.run_or_die(cmd)
+
+            # Not sure why/whether this is needed. But we saw failures w/o it.
+            # We need to investigate how to remove the sleep to speed up
+            time.sleep(5)
+
+
+class ClusterNode:
```

Comment:
It feels like we are essentially wrapping a host instance, and maybe we should do that explicitly (passing in host at construction time) adding all the cluster node specific stuff on top?

Reply:
Well, not exactly.  We only need the `host.Host` for VM nodes.  I'll just pass that to the `VmClusterNode()` constructor.  All the rest don't need an explicit hostconn.  And that kind of makes sense, baremetal (x86 or BF2) get reprovisioned from scratch so why would they need a hostconn.